### PR TITLE
Add support for checkpoint/restore into and out of pods

### PR DESCRIPTION
**Pull request by adrianreber**
This adds support to checkpoint containers out of pods and restore container into pods.

It is only possible to restore a container into a pod if it has been checkpointed out of pod. It is also not possible to restore a non pod container into a pod.

The main reason this does not work is the PID namespace. If a non pod container is being restored in a pod with a shared PID namespace, at least one process in the restored container uses PID 1 which is already in use by the infrastructure container. If someone tries to restore a container from a pod with a shared PID namespace without a shared PID namespace it will also fail because the resulting PID namespace will not have a PID 1.

### CODE REVIEW

File: libpod/container_internal.go
Comment (mheon) [inline libpod/container_internal.go:440]: Extra debug?

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:908]: First letter should not be capitalized

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:919]: First letter should not be capitalized

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:915]: Need to lock the pod before this

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:918]: Probably easier to drop the updatePod and instead just call `pod.InfraContainerID()` to get the ID

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:1291]: Need to lock the infra container before this.

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:1300]: This isn't really matching the way we do pods - the pod itself stores what namespaces it shares. I would not expect good results if you restore a container from a pod that shared the PID namespace, into a pod that does not share the PID namespace.

Comment (mheon) [reply libpod/container_internal_linux.go:1300]: Also, for all of these - you need to ensure that the container config field is set to the ID of the new pod's infra container. I think that the DB will not allow you to add a container to the DB if this isn't true, but we should make sure.

File: libpod/oci_conmon_linux.go
Comment (mheon) [inline libpod/oci_conmon_linux.go:1078]: Need to lock the infra container before doing this

Comment (mheon) [reply libpod/oci_conmon_linux.go:1078]: Actually, do we need to do this at all? The Mount and Process label should not require a sync

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:1278]: We need to make sure that the container config for the restored container has the ID of the new pod in it.

**Comment by mheon** [comment]
I'm somewhat doubtful that checkpointing a container and restoring it into a different pod will work well - there's the potential for configuration differences that the container itself can't really detect but would result in major issues (a missing shared namespace being most notable)

File: libpod/container_internal_linux.go
Comment (adrianreber) [reply libpod/container_internal_linux.go:1304]: I added checks to make sure only container are restored into Pods with the same namespace shared as the checkpointed container.

File: libpod/container_internal_linux.go
Comment (adrianreber) [reply libpod/container_internal_linux.go:1279]: I think I solved also this.

**Comment by adrianreber** [comment]
@mheon Thanks for your review. I added a check to only restore containers into a pod that has the same namespaces shared as the original pod out of which the container was checkpointed. I also updated all container metadata to no longer reference the original pod, but updated it with the new pod.

**Comment by rhatdan** [comment]
Should we have a podman pod checkpoint and podman pod restore?

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:1297]: This should move down a bit, just under the `initAndStart()` on the infra container

**Comment by adrianreber** [comment]
> Should we have a podman pod checkpoint and podman pod restore?

Yes, that was my goal. This is just the first step towards that.

File: libpod/container_internal_linux.go
Comment (adrianreber) [reply libpod/container_internal_linux.go]: done

File: test/e2e/checkpoint_test.go
Comment (edsantiago) [inline test/e2e/checkpoint_test.go:1015]: This pattern is anti-recommended, please see #10997. Could you fix it (and all instances) to use the recommended form? Thank you!

**Review by edsantiago — CHANGES_REQUESTED** [review]
Recommended form is `Expect(session).To(Exit(0))`

File: test/e2e/checkpoint_test.go
Comment (adrianreber) [reply test/e2e/checkpoint_test.go:1015]: done

File: libpod/oci_conmon_linux.go
Comment (adrianreber) [reply libpod/oci_conmon_linux.go]: I simplified this part.

**Comment by vrothberg** [comment]
> Recommended form is `Expect(session).To(Exit(0))`

That's still open, after that we can merge.

**Comment by adrianreber** [comment]
> > Recommended form is `Expect(session).To(Exit(0))`
> 
> That's still open, after that we can merge.

Oops, I thought I changed it. I will take a look later today.

**Comment by adrianreber** [comment]
> > > Recommended form is `Expect(session).To(Exit(0))`
> > 
> > 
> > That's still open, after that we can merge.
> 
> Oops, I thought I changed it. I will take a look later today.

I pushed but did not commit previously. But now I committed and pushed.

**Review by edsantiago — APPROVED** [review]

**Review by vrothberg — COMMENTED** [review]
LGTM, thank you @adrianreber!

@containers/podman-maintainers PTAL

**Pull request merged**
